### PR TITLE
[GEOS-9196] decode special characters with relative paths

### PR DESCRIPTION
**Pull request by NielsCharlier**
https://osgeo-org.atlassian.net/browse/GEOS-9196

<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by aaime** [comment]
Looks good to me... @jodygarnett could you have a second look? Resource is kind of your area of expertise.

File: src/platform/src/test/java/org/geoserver/platform/GeoServerResourceLoaderTest.java [hunk 1]
```diff
@@ -162,4 +166,22 @@ public void testSetBaseDirectory() throws IOException {
...
+
+    @Test
+    public void fromRelativeURLTest() throws IOException {
+        GeoServerResourceLoader loader = new GeoServerResourceLoader();
+        tempFolder.create();
+        File tempDir = tempFolder.getRoot();
+        loader.setBaseDirectory(tempDir);
+
+        Resource res = loader.fromURL("file:relative/with/special+characters%23");
+        assertEquals("relative/with/special characters#", res.path());
```
Comment (jodygarnett) [inline src/platform/src/test/java/org/geoserver/platform/GeoServerResourceLoaderTest.java:178]: I always get frustrated with special characters.

@NielsCharlier can you add a test for the unicode precomposed characters which are known to cause problems on macOS.

> For example, the character "ü" can be encoded as the single code point U+00FC "ü" or as the base character U+0075 "u" followed by the non-spacing character U+0308 "¨"

File: src/platform/src/main/java/org/geoserver/platform/resource/Resources.java [hunk 2]
```diff
@@ -593,6 +595,12 @@ public static Resource fromURL(Resource baseDirectory, String url) {
         if (url.startsWith("file:")) {
             url = url.substring(5); // remove 'file:' prefix
 
+            // revert encoded special characters and spaces
+            try {
+                url = URLDecoder.decode(url, "UTF-8");
+            } catch (UnsupportedEncodingException e) {
```
Comment (jodygarnett) [inline src/platform/src/main/java/org/geoserver/platform/resource/Resources.java:601]: Why is this exception being eaten? nom nom nom ...

Comment (jodygarnett) [reply src/platform/src/main/java/org/geoserver/platform/resource/Resources.java:601]: Q: This is detecting for relative paths, but misses the absolute path thing above?
Q: Does URLs.urlToFile need similar logic?

Comment (NielsCharlier) [reply src/platform/src/main/java/org/geoserver/platform/resource/Resources.java:601]: The exception is eaten, because it cannot occur.
See https://docs.oracle.com/javase/8/docs/api/java/nio/charset/Charset.html
"Every implementation of the Java platform is required to support the following standard charsets. "

As explained on the jira ticket, absolute paths already support special characters, and only relative ones fail. This is exactly because URLS.urlTOFile already has the logic, and it is being called for absolute paths.

File: src/platform/src/test/java/org/geoserver/platform/GeoServerResourceLoaderTest.java (hunk 1 above)
Comment (NielsCharlier) [reply src/platform/src/test/java/org/geoserver/platform/GeoServerResourceLoaderTest.java]: DOne.

**Comment by aaime** [comment]
Looks like one of @jodygarnett questions is still outstanding

**Comment by NielsCharlier** [comment]
Which one? He asked the same questions twice, and they are answered above...

**Comment by aaime** [comment]
Ah ok, saw it now... yeah, twice the same question. Then I think we just need @jodygarnett to complete the review.

File: src/platform/src/test/java/org/geoserver/platform/GeoServerResourceLoaderTest.java (hunk 1 above)
Comment (jodygarnett) [reply src/platform/src/test/java/org/geoserver/platform/GeoServerResourceLoaderTest.java]: Thanks, now to check if the macOS build passes :)

**Comment by jodygarnett** [comment]
Thanks Niels, rebased onto master, backport if needed.

**Comment by NielsCharlier** [comment]
Thanks @jodygarnett
